Python implementation of Z-algorithm

modelflat opened the pull request:
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

AC proof: https://judge.yosupo.jp/submission/239873

SansPapyrus683 commented on content/6_Advanced/String_Search.mdx:
```diff
@@ -197,6 +197,27 @@ int main() {
+from typing import List
+
+
+def z_function(s: str) -> List[int]:
```
this seems to do something different from the c++ code?

TheGamingMousse replied:
honestly i think it's better to have the z function code separate from the code that handles input n stuff

but yeah this is true

modelflat replied:
@SansPapyrus683, yep, it's different. I think my version is better cause it's isolated from any specific problem and I/O stuff. Makes it easier to understand and reuse. We should probably refactor the C++ version in a similar fashion. What do you think?

SansPapyrus683 replied:
i agree, though this impl is supposed to be used to solve that "finding periods" problem above

modelflat replied:
I updated it to solve finding periods. Also extracted Z-algo in C++ implementation into a separate function. Let me know what you think.

AC C++ https://cses.fi/problemset/result/10765731/
AC Python https://cses.fi/problemset/result/10765787/

SansPapyrus683 commented on content/6_Advanced/String_Search.mdx:
```diff
@@ -188,15 +186,58 @@ int main() {
+int main() {
+	string s;
+	cin >> s;
+	const auto z = z_function(s);
```
plz no auto

SansPapyrus683 commented on content/6_Advanced/String_Search.mdx:
```diff
@@ -188,15 +186,58 @@ int main() {
+	return z
+
+
+def main():
```
is this needed

SansPapyrus683 replied:
like putting the code inside a function instead of the file directly

modelflat replied:
I prefer `main` to putting stuff in global scope because it has a local scope, thus avoiding clashes between variables. However I think for the brevity purposes we can get rid of `main`. Updated.

TheGamingMousse commented on content/6_Advanced/String_Search.mdx:
```diff
@@ -175,9 +175,7 @@ We would then have these two cases:
-int main() {
-	string s;
-	cin >> s;
+vector<int> z_function(const string &s) {
```
are we defining `z[0]` as being `n` or `0`? cuz the python one defines it as `n` (which I think makes more sense, IMO)

TheGamingMousse commented on content/6_Advanced/String_Search.mdx:
```diff
@@ -175,9 +175,7 @@ We would then have these two cases:
+vector<int> z_function(const string &s) {
 	vector<int> z(s.size());
 
 	for (int i = 1, l = 0, r = 0; i < s.size(); i++) {
```
i think it would look nicer if the `l = 0` and `r = 0` came before the loop, like how it is in the python version

SansPapyrus683 replied:
![](https://media1.tenor.com/m/6rmxp2-dNrcAAAAC/class-of-09-this-is-true.gif)

modelflat replied:
Fixed!

modelflat commented:
Addressed comments, thanks @SansPapyrus683 @TheGamingMousse !

AC Python updated https://cses.fi/problemset/result/10772986/
AC C++ updated https://cses.fi/problemset/result/10773040/

SansPapyrus683 commented:
yeah np
btw when you're done you can click the "resolve convo" button at the bottom

SansPapyrus683 commented on content/6_Advanced/String_Search.mdx:
```diff
@@ -175,28 +175,67 @@ We would then have these two cases:
+	l, r = 0, 0
+	for i in range(1, n):
+		zi = min(r - i, z[i - l]) if i < r else 0
+		while i + zi < n and s[i + zi] == s[zi]:
+			zi += 1
+		if i + zi > r:
+			l, r = i, i + zi
```
noticing a small diff in impl between this while loop & the one in the c++ version
